src/main/scala/play-json.scala

```diff
@@ -444,6 +444,82 @@ This can be caused by https://issues.scala-lang.org/browse/SI-7046 which can onl
     t
   }
 
+  def formatHinted[T: c.WeakTypeTag](enclosed: Tree): Tree = {
+    assertClass[T]()
+    val T = c.weakTypeOf[T]
+
+    val t =
+      q"""
+      {
+        import $pjson._
+        import $pkg._
+        new Format[$T] {
+          private val _delegateFormat: Format[$T] = $enclosed
+          private val _hints: Hints = implicitly[Hints]
+
+          def reads(json: JsValue): JsResult[$T] = {
+            (json \ _hints.field).asOpt[String] match {
+              case Some(hint) if _hints.isHintForClass(hint, classOf[$T]) => _delegateFormat.reads(json)
+              case Some(unexpectedHint) => JsError(s"Cannot deserialize type [$${classOf[$T]}] from json with type-hint [$${unexpectedHint}]")
+              case None => JsError(s"Cannot deserialize type [$${classOf[$T]}] from json without a type-hint field [_hints.field]")
+            }
+          }
+
+          def writes(value: $T): JsValue = _delegateFormat.writes(value) match {
+            case obj: JsObject => obj + (_hints.field, JsString(_hints.hintFor(classOf[$T])))
+            case nonObj => throw new Exception(s"Cannot put type-hint to $${nonObj.getClass.getSimpleName} produced by Format["+${Literal(Constant(T.toString))}+"]. Hinting supported only for Formats that write JsObjects")
+          }
+        }
+      }
+      """
+//    debugMacro(t)
+    t
+  }
+
+  def formatHintedSealed[T: c.WeakTypeTag]: Tree = {
+    assertSealedAbstract[T]
+
+    val T = c.weakTypeOf[T]
+    val subs = T.typeSymbol.asClass.knownDirectSubclasses.toVector
+    if (subs.isEmpty)
+      c.error(c.enclosingPosition,s"""
+No child classes found for $T. If there clearly are child classes,
+try moving the call into a separate file, a sibbling package, a separate sbt sub project or else.
+This can be caused by https://issues.scala-lang.org/browse/SI-7046 which can only be avoided by manually moving the call.""")
```

Comment:
How about we start a Wiki page about it and link to that instead? I also had opposite cases where just moving it lower in the file helped and a separate package was not needed.

Oh also, could you add a documentation section in the readme about formatHinted? I think it is a good addition. What I like about it in particular is, that formatSealed is actually dangerous because if you have `case class Foo(i: Int)` and `case class Bar(i: Int, s:String)` and you try to deserialize `{i: 5, s:"baz"}` it may end up as Foo, because that works and order is not guaranteed. formatHintedSealed does not have that problem.

How about replacing the word hint with tag? It's not really a hint, it is actually a mandatory type and literature seems to call this a tag in this context (See https://en.wikipedia.org/wiki/Tagged_union). So maybe  formatHinted => formatTagged, formatHintedSealed => formatSealedTagged (also switched the order there), Hint => Tag